include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,163 @@
+#include <unifex/get_stop_token.hpp>
+#include <unifex/async_trace.hpp>
+
+#include <functional>
+#include <type_traits>
+
+namespace execution {
+class sequenced_policy;
+class parallel_policy;
+}
+
+namespace unifex {
+
+template <typename Predecessor, typename Policy, typename Range, typename Func>
+struct indexed_for_sender {
+  UNIFEX_NO_UNIQUE_ADDRESS Predecessor pred_;
+  UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+  UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template <template <typename...> class Tuple>
+  struct indexed_for_result {
+   public:
+    template <typename... Args>
+    using apply = Tuple<Args...>;
+  };
+
+  template<typename... Args>
+  using is_overload_noexcept = std::is_nothrow_invocable<Func, Args...>;
+
+  template <template <typename...> class Variant>
+  struct calculate_errors {
+   public:
+    template <typename... Errors>
+    using apply = deduplicate<Variant<Errors..., std::exception_ptr>>;
+  };
+
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = deduplicate_t<typename Predecessor::template value_types<
+      Variant,
+      indexed_for_result<Tuple>::template apply>>;
+
+  template <template <typename...> class Variant>
+  using error_types = typename Predecessor::template error_types<
+      calculate_errors<Variant>::template apply>;
+
+  friend constexpr auto tag_invoke(
+      tag_t<blocking>,
+      const indexed_for_sender& sender) {
+    return blocking(sender.pred_);
+  }
+
+  template <typename Receiver>
+  struct indexed_for_receiver {
+    UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+    UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+    UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+    // sequenced_policy version supports forward range
+    template<typename... Values>
+    static void apply_func_with_policy(const execution::sequenced_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(std::is_nothrow_invocable_v<Func, typename std::iterator_traits<typename Range::iterator>::reference, Values...>) {
+      for(auto idx : range) {
+        std::invoke(func, idx, values...);
+      }
+    }
+
+    // parallel_policy version requires random access range
+    template<typename... Values>
+    static void apply_func_with_policy(const execution::parallel_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(std::is_nothrow_invocable_v<Func, typename std::iterator_traits<typename Range::iterator>::reference, Values...>) {
+      auto start = range.begin();
+      for(auto idx = 0; idx < range.size(); ++idx) {
+        std::invoke(func, start[idx], values...);
+      }
+    }
+
+    template <typename... Values>
+    void set_value(Values&&... values) && noexcept {
+      if constexpr (std::is_nothrow_invocable_v<Func, typename std::iterator_traits<typename Range::iterator>::reference, Values...>) {
```

Comment:
Should be `std::is_nothrow_invocable_v<Func&, ...>` (note the lvalue-reference).